# Quiz &mdash; Files

For the following exercises, assume that the file `pi_digits.txt` is under `assets` under the current directory and contains 3 lines as follows:

In [1]:
!cat assets/pi_digits.txt

3.1415926535
  8979323846
  2643383279

A. Given

In [11]:
with open('assets/pi_digits.txt', mode='r') as f:
    pi_txt = f.read()
    
pi_txt

'3.1415926535\n  8979323846\n  2643383279'

Which is correct (check all that apply):

1. [ ] On the first line, `as f` must be `as file_obj`
2. [ ] On the last line, `pi_txt` is out of scope
2. [ ] Prints π with 31 digits
2. [ ] Prints the contents of the file
2. [ ] Prints the first line of the file followed by a blank line
2. [ ] A `FileNotFound` exception is raised

B. Given

In [3]:
with open('assets/pi_digits.txt', mode='r') as f:
    pi_txt = ''
    line = f.readline()
    while line:
        pi_txt += line.strip()
        line = f.readline()

print(pi_txt)

3.141592653589793238462643383279


Which is correct (check all that apply):

1. [ ] In the last line, `pi_txt` is out of scope
1. [ ] Prints `3.141592653589793238462643383279`
1. [ ] Prints the 3 lines of the file, but with no leading spaces
1. [ ] There is an error in the code

C. Given

In [4]:
with open('assets/pi_digits.txt', mode='r') as f:
    pi_txt = ''
    while line = f.readline():
        pi_txt += line.strip()
    print('π', '=', pi_txt[:4], '...', pi_txt[-3:])

Which is correct (check all that apply):

1. [ ] Prints `π=3.14...279`
1. [ ] Prints: `π = 3.14 ... 279`
1. [ ] The print statement has too many arguments
1. [ ] You have to use unicode escape to print special symbols such as π, i.e., `print('\u03C0', ...)`
1. [ ] The print statement does not need to be indented
1. [ ] There's an error in the code

In [5]:
print('\u03C0')

π


D. Given

In [6]:
with open('assets/pi_digits.txt', mode='r') as f:
    pi_txt = ''.join(f.readlines())
print('...'.join((pi_txt[:4], pi_txt[-3:])))

3.14...279


Which is correct (check all that apply):
1. [ ] Prints `3.14...79` and a blank line
1. [ ] Prints: `3.14 ... 279`
1. [ ] There's a typo: `readlines()` should be `readline()` 
1. [ ] The print statement must be indented
1. [ ] There are too many parentheses in the print statement

E. Given

In [7]:
with open('assets/pi_digits.txt', mode='r') as f:
    pi_txt = ''.join(line.strip() for line in f)
print(pi_txt[:4], '...', pi_txt[-3:], sep='')

3.14...279


Which is correct (check all that apply):

1. [ ] Prints `3.14...279`
1. [ ] There's a missing `readlines()` on the second line
1. [ ] The print statement could have been indented
1. [ ] `line.strip()` should have been `line.rstrip()`

F. Given

In [8]:
with open('assets/pi_digits.txt', mode='r') as f:
    lines = f.readlines()
pi_txt = ''.join(line.strip() for line in lines)
print(pi_txt[:4], pi_txt[-3:], sep='...')

3.14...279


Which is correct (check all that apply):

1. [ ] Prints `3.14...279`
1. [ ] Third line needs to be indented
1. [ ] The print statement could have been indented
1. [ ] `line.strip()` should have been `line.rstrip()`

G. Given

In [9]:
with open('assets/pi_digits.txt', mode='r') as f:
    lines = f
pi_txt = ''.join(line.strip() for line in lines)
print(pi_txt[:4], '.', pi_txt[-3:], sep='.')

Which is correct (check all that apply):

1. [ ] Prints `3.14...279`
1. [ ] Third line needs to be indented
1. [ ] The print statement could have been indented
1. [ ] There's an error in the code

H. Given

In [10]:
with open('assets/pi_digits.txt', mode='r') as f:
    head = next(f).strip()
    for line in f: 
        pass
    tail = line.strip()
print('...'.join((head[:4], tail[-3:])))

3.14...279


Which is correct (check all that apply):

1. [ ] Prints `3.14...279`
1. [ ] `pass` is just a placeholder, which needs to be replaced with proper code before execution.
1. [ ] `line.strip()` will raise an exception because `line` is `None` after exiting the loop
1. [ ] `line.strip()` will raise an exception because `line` is out of scope after exiting the loop
1. [ ] You cannot call `next(f)` and then loop over `f`